File: shortcut_analyzer.cpp
// shortcut_analyzer.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <fstream>
#include <sstream>
#include <algorithm>
#include <cctype>
#include <json/json.h> // sudo apt-get install libjsoncpp-dev

using namespace std;

const string RESET = "\033[0m";
const string GREEN = "\033[92m";
const string RED = "\033[91m";
const string YELLOW = "\033[93m";
const string BLUE = "\033[94m";
const string CYAN = "\033[96m";
const string BOLD = "\033[1m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

struct Shortcut {
    string name;
    string shortcut;
    string app;
    string category;
};

class Analyzer {
public:
    vector<Shortcut> shortcuts;
    map<string, vector<Shortcut>> apps;
    map<string, vector<Shortcut>> categories;
    map<string, vector<Shortcut>> conflicts;

    Analyzer(const Json::Value& root) {
        auto& arr = root["shortcuts"];
        for (const auto& item : arr) {
            Shortcut s;
            s.name = item["name"].asString();
            s.shortcut = item["shortcut"].asString();
            s.app = item["app"].asString();
            s.category = item["category"].asString();
            shortcuts.push_back(s);
        }
        index();
    }

    void index() {
        for (auto& s : shortcuts) {
            string app = s.app.empty() ? "Без приложения" : s.app;
            apps[app].push_back(s);

            string cat = s.category.empty() ? "Без категории" : s.category;
            categories[cat].push_back(s);

            if (!s.shortcut.empty()) {
                conflicts[s.shortcut].push_back(s);
            }
        }
    }

    vector<Shortcut> search(const string& query) {
        vector<Shortcut> results;
        string q = query;
        transform(q.begin(), q.end(), q.begin(), ::tolower);
        for (auto& s : shortcuts) {
            string name = s.name;
            transform(name.begin(), name.end(), name.begin(), ::tolower);
            string sc = s.shortcut;
            transform(sc.begin(), sc.end(), sc.begin(), ::tolower);
            if (name.find(q) != string::npos || sc.find(q) != string::npos) {
                results.push_back(s);
            }
        }
        return results;
    }

    map<string, vector<Shortcut>> getConflicts() {
        map<string, vector<Shortcut>> result;
        for (auto& kv : conflicts) {
            if (kv.second.size() > 1) {
                result[kv.first] = kv.second;
            }
        }
        return result;
    }

    void exportCSV(const string& filename) {
        ofstream f(filename);
        if (!f) { cerr << colorize("Ошибка создания файла", RED) << endl; return; }
        f << "name,shortcut,app,category\n";
        for (auto& s : shortcuts) {
            f << s.name << "," << s.shortcut << "," << s.app << "," << s.category << "\n";
        }
        cout << colorize("Экспортировано в " + filename, GREEN) << endl;
    }

    void display(const vector<Shortcut>* data, bool group, bool conflictsOnly, bool verbose) {
        const vector<Shortcut>* items = data ? data : &shortcuts;
        if (items->empty()) {
            cout << colorize("Нет данных для отображения.", YELLOW) << endl;
            return;
        }
        if (conflictsOnly) {
            auto conf = getConflicts();
            if (conf.empty()) {
                cout << colorize("Конфликтов не найдено.", GREEN) << endl;
                return;
            }
            cout << colorize("🔍 Конфликты (одинаковые комбинации):", BOLD) << endl;
            for (auto& kv : conf) {
                cout << colorize("  " + kv.first + ":", RED) << endl;
                for (auto& s : kv.second) {
                    cout << "    - " << s.name << " (" << s.app << ")" << endl;
                }
            }
            return;
        }
        if (group) {
            for (auto& kv : apps) {
                cout << colorize("\n📁 " + kv.first + " (" + to_string(kv.second.size()) + "):", BLUE) << endl;
                for (auto& s : kv.second) {
                    printItem(s, verbose);
                }
            }
        } else {
            for (auto& s : *items) {
                printItem(s, verbose);
            }
        }
    }

    void printItem(const Shortcut& s, bool verbose) {
        if (verbose) {
            cout << "  " << colorize(s.name, BOLD) << " → " << colorize(s.shortcut, CYAN)
                 << "  (приложение: " << s.app << ", категория: " << s.category << ")" << endl;
        } else {
            cout << "  " << s.name << " → " << s.shortcut << endl;
        }
    }
};

Json::Value loadJSON(const string& filename) {
    ifstream f(filename);
    Json::Value root;
    f >> root;
    return root;
}

int main(int argc, char* argv[]) {
    string file, search, exportFile;
    bool group = false, conflictsOnly = false, verbose = false;

    for (int i=1; i<argc; ++i) {
        string arg = argv[i];
        if (arg == "-s" && i+1 < argc) search = argv[++i];
        else if (arg == "-g") group = true;
        else if (arg == "-c") conflictsOnly = true;
        else if (arg == "-e" && i+1 < argc) exportFile = argv[++i];
        else if (arg == "-v") verbose = true;
        else if (arg == "-h") {
            cout << "Usage: shortcut_analyzer <file> [-s search] [-g] [-c] [-e file] [-v]" << endl;
            return 0;
        } else if (file.empty()) file = arg;
    }
    if (file.empty()) {
        cerr << colorize("Укажите файл конфигурации.", RED) << endl;
        return 1;
    }

    Json::Value root;
    try {
        root = loadJSON(file);
    } catch (const exception& e) {
        cerr << colorize("Ошибка загрузки файла: " + string(e.what()), RED) << endl;
        return 1;
    }

    Analyzer analyzer(root);

    if (!search.empty()) {
        auto results = analyzer.search(search);
        if (!results.empty()) {
            cout << colorize("🔍 Найдено " + to_string(results.size()) + " совпадений:", BOLD) << endl;
            analyzer.display(&results, group, false, verbose);
        } else {
            cout << colorize("Ничего не найдено.", YELLOW) << endl;
        }
        return 0;
    }

    if (conflictsOnly) {
        analyzer.display(nullptr, false, true, false);
        return 0;
    }

    if (!exportFile.empty()) {
        if (exportFile.find(".csv") != string::npos) {
            analyzer.exportCSV(exportFile);
        } else if (exportFile.find(".json") != string::npos) {
            // Для JSON используем простой вывод, в реальном проекте можно улучшить
            ofstream f(exportFile);
            f << root.toStyledString();
            cout << colorize("Экспортировано в " + exportFile, GREEN) << endl;
        } else {
            cerr << colorize("Неизвестный формат экспорта. Используйте .csv или .json", RED) << endl;
        }
        return 0;
    }

    analyzer.display(nullptr, group, false, verbose);
    return 0;
}

File: shortcut_analyzer.go
// shortcut_analyzer.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"strings"
)

const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
	blue   = "\033[94m"
	cyan   = "\033[96m"
	bold   = "\033[1m"
)

func colorize(text, color string) string {
	return color + text + reset
}

type Shortcut struct {
	Name     string `json:"name"`
	Shortcut string `json:"shortcut"`
	App      string `json:"app"`
	Category string `json:"category"`
}

type Data struct {
	Shortcuts []Shortcut `json:"shortcuts"`
}

type Analyzer struct {
	shortcuts []Shortcut
	apps      map[string][]Shortcut
	categories map[string][]Shortcut
	conflicts map[string][]Shortcut
}

func NewAnalyzer(data Data) *Analyzer {
	a := &Analyzer{
		shortcuts: data.Shortcuts,
		apps:      make(map[string][]Shortcut),
		categories: make(map[string][]Shortcut),
		conflicts: make(map[string][]Shortcut),
	}
	a.index()
	return a
}

func (a *Analyzer) index() {
	for _, s := range a.shortcuts {
		app := s.App
		if app == "" {
			app = "Без приложения"
		}
		a.apps[app] = append(a.apps[app], s)

		cat := s.Category
		if cat == "" {
			cat = "Без категории"
		}
		a.categories[cat] = append(a.categories[cat], s)

		if s.Shortcut != "" {
			a.conflicts[s.Shortcut] = append(a.conflicts[s.Shortcut], s)
		}
	}
}

func (a *Analyzer) Search(query string) []Shortcut {
	var results []Shortcut
	q := strings.ToLower(query)
	for _, s := range a.shortcuts {
		if strings.Contains(strings.ToLower(s.Name), q) || strings.Contains(strings.ToLower(s.Shortcut), q) {
			results = append(results, s)
		}
	}
	return results
}

func (a *Analyzer) GetConflicts() map[string][]Shortcut {
	conf := make(map[string][]Shortcut)
	for k, v := range a.conflicts {
		if len(v) > 1 {
			conf[k] = v
		}
	}
	return conf
}

func (a *Analyzer) GroupByApp() map[string][]Shortcut {
	return a.apps
}

func (a *Analyzer) ExportCSV(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()
	writer := csv.NewWriter(file)
	defer writer.Flush()
	headers := []string{"name", "shortcut", "app", "category"}
	writer.Write(headers)
	for _, s := range a.shortcuts {
		writer.Write([]string{s.Name, s.Shortcut, s.App, s.Category})
	}
	fmt.Println(colorize("Экспортировано в "+filename, green))
	return nil
}

func (a *Analyzer) ExportJSON(filename string) error {
	data := Data{Shortcuts: a.shortcuts}
	jsonData, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	return ioutil.WriteFile(filename, jsonData, 0644)
}

func (a *Analyzer) Display(shortcuts []Shortcut, group, conflictsOnly, verbose bool) {
	data := shortcuts
	if data == nil {
		data = a.shortcuts
	}
	if len(data) == 0 {
		fmt.Println(colorize("Нет данных для отображения.", yellow))
		return
	}
	if conflictsOnly {
		conf := a.GetConflicts()
		if len(conf) == 0 {
			fmt.Println(colorize("Конфликтов не найдено.", green))
			return
		}
		fmt.Println(colorize("🔍 Конфликты (одинаковые комбинации):", bold))
		for combo, items := range conf {
			fmt.Printf("%s\n", colorize("  "+combo+":", red))
			for _, item := range items {
				fmt.Printf("    - %s (%s)\n", item.Name, item.App)
			}
		}
		return
	}
	if group {
		grouped := a.GroupByApp()
		for app, items := range grouped {
			fmt.Printf("%s\n", colorize(fmt.Sprintf("\n📁 %s (%d):", app, len(items)), blue))
			for _, item := range items {
				a.printItem(item, verbose)
			}
		}
	} else {
		for _, item := range data {
			a.printItem(item, verbose)
		}
	}
}

func (a *Analyzer) printItem(item Shortcut, verbose bool) {
	if verbose {
		fmt.Printf("  %s → %s  (приложение: %s, категория: %s)\n",
			colorize(item.Name, bold),
			colorize(item.Shortcut, cyan),
			item.App,
			item.Category)
	} else {
		fmt.Printf("  %s → %s\n", item.Name, item.Shortcut)
	}
}

func loadJSON(filename string) (Data, error) {
	var data Data
	bytes, err := ioutil.ReadFile(filename)
	if err != nil {
		return data, err
	}
	err = json.Unmarshal(bytes, &data)
	return data, err
}

func main() {
	var (
		file        string
		search      string
		group       bool
		conflicts   bool
		exportFile  string
		verbose     bool
	)
	flag.StringVar(&file, "f", "", "Файл конфигурации (JSON)")
	flag.StringVar(&search, "s", "", "Поиск по названию или комбинации")
	flag.BoolVar(&group, "g", false, "Группировать по приложениям")
	flag.BoolVar(&conflicts, "c", false, "Показать конфликты")
	flag.StringVar(&exportFile, "e", "", "Экспортировать в файл (CSV или JSON)")
	flag.BoolVar(&verbose, "v", false, "Подробный вывод")
	flag.Parse()

	if file == "" && flag.NArg() > 0 {
		file = flag.Arg(0)
	}
	if file == "" {
		fmt.Println(colorize("Укажите файл конфигурации.", red))
		flag.Usage()
		os.Exit(1)
	}

	data, err := loadJSON(file)
	if err != nil {
		fmt.Println(colorize("Ошибка загрузки файла: "+err.Error(), red))
		os.Exit(1)
	}

	analyzer := NewAnalyzer(data)

	if search != "" {
		results := analyzer.Search(search)
		if len(results) > 0 {
			fmt.Printf("%s\n", colorize(fmt.Sprintf("🔍 Найдено %d совпадений:", len(results)), bold))
			analyzer.Display(results, group, false, verbose)
		} else {
			fmt.Println(colorize("Ничего не найдено.", yellow))
		}
		return
	}

	if conflicts {
		analyzer.Display(nil, false, true, false)
		return
	}

	if exportFile != "" {
		var err error
		if strings.HasSuffix(exportFile, ".csv") {
			err = analyzer.ExportCSV(exportFile)
		} else if strings.HasSuffix(exportFile, ".json") {
			err = analyzer.ExportJSON(exportFile)
		} else {
			fmt.Println(colorize("Неизвестный формат экспорта. Используйте .csv или .json", red))
			return
		}
		if err != nil {
			fmt.Println(colorize("Ошибка экспорта: "+err.Error(), red))
		}
		return
	}

	analyzer.Display(nil, group, false, verbose)
}
